src/district/server/smart_contracts.cljs

```diff
@@ -238,80 +238,95 @@
-                                                                :event event
-                                                                :from from
-                                                                :to to})
-
-                       (web3-eth/get-past-events contract-instance
-                                                 event
-                                                 {:from-block from
-                                                  :to-block to}
-                                                 (fn [error events]
-                                                   (let [logs (->> events
-                                                                   web3-helpers/js->cljkk
-                                                                   (map (partial enrich-event-log contract contract-instance)))]
-                                                     (async/put! logs-ch {:err error :logs logs}))))
-                       logs-ch))]
-
-    ;; go chan by chan collecting events
-    (go-loop [all-logs []
-              [logs-ch & rest-logs] logs-chans]
-      (if logs-ch
-        (let [{:keys [:err :logs]} (async/<! logs-ch)
-              logs (map #(assoc % :err err) logs)]
-          ;; keep collecting
-          (recur (into all-logs logs) rest-logs))
-
-        ;; no more channels to read, sort and callback
-        (let [sorted-logs (cond->> (sort-by log-order-triplet all-logs)
-
-                            skip-log-indexes (remove (fn [l]
-                                                       (and (= (:block-number l) from-block)
-                                                            (skip-log-indexes [(:transaction-index l) (:log-index l)]))))
-                            true             transform-fn)]
-          (go-loop [logs sorted-logs]
-            (if (seq logs)
-              (do
-                (let [first-log (first logs)]
-
-                  (when (fn? callback)
-                    (doseq [res (try
-                                  (callback (:err first-log) (dissoc first-log :err))
-                                  (catch js/Error e
-                                    (when crash-on-event-fail?
-                                      (log/error "Server crash. Caused by event processing error with :crash-on-event-fail? true. Disable this flag to skip and continue.")
-                                      (.exit js/process 1))))]
-                      ;; if callback returns a promise or chan we block until it resolves
-                      (cond
-                        (satisfies? cljs.core.async.impl.protocols/ReadPort res)
-                        (<! res)
-
-                        (async-helpers/promise? res)
-                        (<! (async-helpers/promise->chan res))))))
-                (recur (rest logs)))
-
-              (when (fn? on-finish)
-                (on-finish sorted-logs)))))))))
+        next-chunk (fn [[from to]]
```

Comment:
I recommend extracting this and other in-line functions from this _let_ and give them descriptive names. I tried to make sense what this method is doing, but it's quite hard. And every developer after you will have to do it if they have to touch this code. It would be good example and benefit many, besides you have this information in your head fresh right now. After couple months I guess it would be hard to remember for you as well.